source/extensions/injected_credentials/common/secret_reader.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include "envoy/secret/secret_manager.h"
+#include "envoy/secret/secret_provider.h"
+
+#include "source/common/config/datasource.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace InjectedCredentials {
+namespace Common {
+
+inline Secret::GenericSecretConfigProviderSharedPtr
+secretsProvider(const envoy::extensions::transport_sockets::tls::v3::SdsSecretConfig& config,
+                Secret::SecretManager& secret_manager,
+                Server::Configuration::TransportSocketFactoryContext& transport_socket_factory,
+                Init::Manager& init_manager) {
+  if (config.has_sds_config()) {
+    return secret_manager.findOrCreateGenericSecretProvider(config.sds_config(), config.name(),
+                                                            transport_socket_factory, init_manager);
+  } else {
+    return secret_manager.findStaticGenericSecretProvider(config.name());
+  }
+}
+
+// Helper class used to fetch secrets (usually from SDS).
+class SecretReader {
+public:
+  virtual ~SecretReader() = default;
+  virtual const std::string& credential() const PURE;
+};
+
+using SecretReaderSharedPtr = std::shared_ptr<SecretReader>;
+
+class SDSSecretReader : public SecretReader {
+public:
+  SDSSecretReader(Secret::GenericSecretConfigProviderSharedPtr secret_provider, Api::Api& api)
+      : update_callback_(readAndWatchSecret(credential_, secret_provider, api)) {}
+
+  const std::string& credential() const override { return credential_; }
+
+private:
+  Envoy::Common::CallbackHandlePtr
+  readAndWatchSecret(std::string& value,
+                     Secret::GenericSecretConfigProviderSharedPtr& secret_provider, Api::Api& api) {
+    const auto* secret = secret_provider->secret();
+    if (secret != nullptr) {
+      value = Config::DataSource::read(secret->secret(), true, api);
+    }
+
+    return secret_provider->addUpdateCallback([secret_provider, &api, &value]() {
```

Comment:
Is this called on main or worker thread?

Reply:
main thread.

Reply:
I this lambda needs to post back to the worker so that `value` is updated on the worker. Otherwise, it's a data race.

Reply:
Can you respond here, please? I think you have a data race here because `credential_` is created on a worker thread but the update callback modifies it on the main thread. You probably either need to capture the worker event dispatcher and post there the update, or use thread local storage.

Reply:
> Can you respond here, please? I think you have a data race here because `credential_` is created on a worker thread but the update callback modifies it on the main thread. You probably either need to capture the worker event dispatcher and post there the update, or use thread local storage.

@kyessenov This is not a new way to get data from a secret, the same pattern is used in multiple places in the codebase. 
OAuth2 : https://github.com/envoyproxy/envoy/blob/bbc339afe2d65b40d3bacd55d242653d02df1bb5/source/extensions/filters/http/oauth2/filter.h#L59-L74
sxg : https://github.com/envoyproxy/envoy/blob/bbc339afe2d65b40d3bacd55d242653d02df1bb5/contrib/sxg/filters/http/source/filter_config.h#L50-L66

I can't find a good way to refactor this code right now and this PR already took us a long time to get to this point. Would you be okay with merging this PR as is and We can open a follow-up issue to try to solve this problem globally? It will also be eaiser to allow other maintainers and contributors to help with it.